File: code/packages/rust/logic-gates/tests/sequential_test.rs
//! Integration tests for the sequential logic module.
//!
//! Tests cover SR latch, D latch, D flip-flop, register, shift register,
//! and counter — verifying correct state transitions, edge-triggered
//! behavior, and multi-bit storage.

use logic_gates::sequential::*;

// ===========================================================================
// SR Latch tests
// ===========================================================================

#[test]
fn test_sr_latch_set() {
    let (q, qb) = sr_latch(1, 0, 0, 1);
    assert_eq!((q, qb), (1, 0), "Set should produce Q=1, Q_bar=0");
}

#[test]
fn test_sr_latch_reset() {
    let (q, qb) = sr_latch(0, 1, 1, 0);
    assert_eq!((q, qb), (0, 1), "Reset should produce Q=0, Q_bar=1");
}

#[test]
fn test_sr_latch_hold_after_set() {
    let (q, qb) = sr_latch(1, 0, 0, 1);
    let (q, qb) = sr_latch(0, 0, q, qb);
    assert_eq!((q, qb), (1, 0), "Hold after set should keep Q=1");
}

#[test]
fn test_sr_latch_hold_after_reset() {
    let (q, qb) = sr_latch(0, 1, 0, 1);
    let (q, qb) = sr_latch(0, 0, q, qb);
    assert_eq!((q, qb), (0, 1), "Hold after reset should keep Q=0");
}

#[test]
fn test_sr_latch_invalid_both_set() {
    let (q, qb) = sr_latch(1, 1, 0, 1);
    assert_eq!((q, qb), (0, 0), "S=R=1 is invalid, both outputs forced to 0");
}

#[test]
fn test_sr_latch_set_reset_set_sequence() {
    // Set -> Reset -> Set
    let (q, qb) = sr_latch(1, 0, 0, 1);
    assert_eq!((q, qb), (1, 0));
    let (q, qb) = sr_latch(0, 1, q, qb);
    assert_eq!((q, qb), (0, 1));
    let (q, qb) = sr_latch(1, 0, q, qb);
    assert_eq!((q, qb), (1, 0));
}

#[test]
fn test_sr_latch_default_initial_state() {
    // Default Q=0, Q_bar=1; hold should preserve
    let (q, qb) = sr_latch(0, 0, 0, 1);
    assert_eq!((q, qb), (0, 1));
}

// ===========================================================================
// D Latch tests
// ===========================================================================

#[test]
fn test_d_latch_store_one() {
    let (q, qb) = d_latch(1, 1, 0, 1);
    assert_eq!((q, qb), (1, 0));
}

#[test]
fn test_d_latch_store_zero() {
    let (q, qb) = d_latch(0, 1, 0, 1);
    assert_eq!((q, qb), (0, 1));
}

#[test]
fn test_d_latch_hold_when_disabled() {
    let (q, qb) = d_latch(1, 1, 0, 1); // Store 1
    let (q, qb) = d_latch(0, 0, q, qb); // Disable, try to store 0
    assert_eq!((q, qb), (1, 0), "Should hold 1 when enable=0");
}

#[test]
fn test_d_latch_transparent_when_enabled() {
    let (q, qb) = d_latch(1, 1, 0, 1);
    assert_eq!(q, 1);
    let (q, _qb) = d_latch(0, 1, q, qb); // Still enabled, change data
    assert_eq!(q, 0, "Should follow data when enable=1");
}

#[test]
fn test_d_latch_full_sequence() {
    // Store 1, hold, store 0, hold
    let (q, qb) = d_latch(1, 1, 0, 1);
    assert_eq!(q, 1);
    let (q, qb) = d_latch(1, 0, q, qb); // hold
    assert_eq!(q, 1);
    let (q, qb) = d_latch(0, 1, q, qb); // store 0
    assert_eq!(q, 0);
    let (q, _qb) = d_latch(1, 0, q, qb); // hold (data=1 ignored)
    assert_eq!(q, 0);
}

// ===========================================================================
// D Flip-Flop tests
// ===========================================================================

#[test]
fn test_d_flip_flop_captures_on_rising_edge() {
    let mut state = FlipFlopState::default();
    // Clock low: master absorbs data=1
    d_flip_flop(1, 0, &mut state);
    // Clock high: slave outputs
    let (q, _qb) = d_flip_flop(1, 1, &mut state);
    assert_eq!(q, 1, "Should capture 1 on rising edge");
}

#[test]
fn test_d_flip_flop_holds_on_clock_high() {
    let mut state = FlipFlopState::default();
    // Rising edge with data=1
    d_flip_flop(1, 0, &mut state);
    d_flip_flop(1, 1, &mut state);
    // Now change data to 0 while clock stays high
    let (q, _qb) = d_flip_flop(0, 1, &mut state);
    // Master is holding (clock=1 means NOT(clock)=0), so data change ignored
    assert_eq!(q, 1, "Should hold value while clock is high");
}

#[test]
fn test_d_flip_flop_captures_zero() {
    let mut state = FlipFlopState::default();
    // First store a 1
    d_flip_flop(1, 0, &mut state);
    d_flip_flop(1, 1, &mut state);
    // Now store a 0 on next rising edge
    d_flip_flop(0, 0, &mut state);
    let (q, _qb) = d_flip_flop(0, 1, &mut state);
    assert_eq!(q, 0, "Should capture 0 on second rising edge");
}

#[test]
fn test_d_flip_flop_multiple_cycles() {
    let mut state = FlipFlopState::default();
    let data_sequence = [1, 0, 1, 1, 0];
    for &data in &data_sequence {
        d_flip_flop(data, 0, &mut state);
        let (q, _qb) = d_flip_flop(data, 1, &mut state);
        assert_eq!(q, data, "Flip-flop should capture data={data}");
    }
}

// ===========================================================================
// Register tests
// ===========================================================================

#[test]
fn test_register_4bit() {
    let mut state: Vec<FlipFlopState> =
        (0..4).map(|_| FlipFlopState::default()).collect();
    register(&[1, 0, 1, 1], 0, &mut state);
    let out = register(&[1, 0, 1, 1], 1, &mut state);
    assert_eq!(out, vec![1, 0, 1, 1]);
}

#[test]
fn test_register_8bit() {
    let mut state: Vec<FlipFlopState> =
        (0..8).map(|_| FlipFlopState::default()).collect();
    let data = vec![1, 0, 1, 0, 1, 0, 1, 0];
    register(&data, 0, &mut state);
    let out = register(&data, 1, &mut state);
    assert_eq!(out, data);
}

#[test]
fn test_register_holds_value() {
    let mut state: Vec<FlipFlopState> =
        (0..4).map(|_| FlipFlopState::default()).collect();
    // Store [1, 1, 0, 0]
    register(&[1, 1, 0, 0], 0, &mut state);
    register(&[1, 1, 0, 0], 1, &mut state);
    // Try to overwrite with [0, 0, 1, 1] but only clock low
    register(&[0, 0, 1, 1], 0, &mut state);
    // Clock high with original data gone — slave should still have [1, 1, 0, 0]
    // Wait, we need to clock high again to see new data
    let out = register(&[0, 0, 1, 1], 1, &mut state);
    assert_eq!(out, vec![0, 0, 1, 1], "Should store new data on next rising edge");
}

#[test]
fn test_register_overwrites() {
    let mut state: Vec<FlipFlopState> =
        (0..4).map(|_| FlipFlopState::default()).collect();
    // First write
    register(&[1, 1, 1, 1], 0, &mut state);
    register(&[1, 1, 1, 1], 1, &mut state);
    // Second write, different data
    register(&[0, 1, 0, 1], 0, &mut state);
    let out = register(&[0, 1, 0, 1], 1, &mut state);
    assert_eq!(out, vec![0, 1, 0, 1]);
}

#[test]
#[should_panic]
fn test_register_panics_on_length_mismatch() {
    let mut state: Vec<FlipFlopState> =
        (0..4).map(|_| FlipFlopState::default()).collect();
    register(&[1, 0, 1], 0, &mut state); // 3 bits for 4-bit register
}

#[test]
#[should_panic]
fn test_register_panics_on_empty_data() {
    let mut state: Vec<FlipFlopState> = vec![];
    register(&[], 0, &mut state);
}

// ===========================================================================
// Shift Register tests
// ===========================================================================

#[test]
fn test_shift_register_right_shift_in_ones() {
    let mut state: Vec<FlipFlopState> =
        (0..4).map(|_| FlipFlopState::default()).collect();

    // Shift in 1 three times
    for i in 0..3 {
        shift_register(1, 0, &mut state, "right");
        let (out, _sout) = shift_register(1, 1, &mut state, "right");
        // After i+1 shifts, first i+1 positions should be 1
        let expected: Vec<u8> = (0..4).map(|j| if j <= i { 1 } else { 0 }).collect();
        assert_eq!(out, expected, "After shift {}", i + 1);
    }
}

#[test]
fn test_shift_register_left_shift() {
    let mut state: Vec<FlipFlopState> =
        (0..4).map(|_| FlipFlopState::default()).collect();

    // Shift in 1 from the right (left direction means entering at high index)
    shift_register(1, 0, &mut state, "left");
    let (out, _sout) = shift_register(1, 1, &mut state, "left");
    assert_eq!(out, vec![0, 0, 0, 1], "Left shift: 1 enters at position 3");
}

#[test]
fn test_shift_register_serial_out() {
    let mut state: Vec<FlipFlopState> =
        (0..4).map(|_| FlipFlopState::default()).collect();

    // Fill register with 1s
    for _ in 0..4 {
        shift_register(1, 0, &mut state, "right");
        shift_register(1, 1, &mut state, "right");
    }

    // Now shift in 0; the serial_out should be 1 (old MSB)
    shift_register(0, 0, &mut state, "right");
    let (_out, sout) = shift_register(0, 1, &mut state, "right");
    assert_eq!(sout, 1, "Serial out should be the old MSB value");
}

#[test]
fn test_shift_register_right_pattern() {
    // Shift in the pattern 1, 0, 1, 0
    let mut state: Vec<FlipFlopState> =
        (0..4).map(|_| FlipFlopState::default()).collect();

    let pattern = [1u8, 0, 1, 0];
    for &bit in &pattern {
        shift_register(bit, 0, &mut state, "right");
        shift_register(bit, 1, &mut state, "right");
    }

    // After shifting in [1, 0, 1, 0], the register should contain [0, 1, 0, 1]
    // because each new bit pushes from position 0
    let (_out, _) = shift_register(0, 0, &mut state, "right");
    // Read current state from slave_q
    let current: Vec<u8> = state.iter().map(|s| s.slave_q).collect();
    assert_eq!(current, vec![0, 1, 0, 1]);
}

#[test]
#[should_panic]
fn test_shift_register_panics_on_bad_direction() {
    let mut state: Vec<FlipFlopState> =
        (0..4).map(|_| FlipFlopState::default()).collect();
    shift_register(0, 0, &mut state, "up");
}

// ===========================================================================
// Counter tests
// ===========================================================================

#[test]
fn test_counter_counts_to_three() {
    let mut state = CounterState::new(4);

    // Count 0 -> 1 -> 2 -> 3
    // counter(clock, reset, state) — clock cycles low->high, reset=0
    let expected = [
        vec![1, 0, 0, 0], // 1
        vec![0, 1, 0, 0], // 2
        vec![1, 1, 0, 0], // 3
    ];

    for (i, exp) in expected.iter().enumerate() {
        counter(0, 0, &mut state); // clock low
        let bits = counter(1, 0, &mut state); // clock high
        assert_eq!(&bits, exp, "Count step {} should be {:?}", i + 1, exp);
    }
}

#[test]
fn test_counter_overflow() {
    // 4-bit counter: max value = 15, then wraps to 0
    let mut state = CounterState::new(4);

    // Count to 15
    for _ in 0..15 {
        counter(0, 0, &mut state);
        counter(1, 0, &mut state);
    }
    // At this point state.value should be [1, 1, 1, 1] = 15
    assert_eq!(state.value, vec![1, 1, 1, 1]);

    // One more tick: overflow to 0
    counter(0, 0, &mut state);
    let bits = counter(1, 0, &mut state);
    assert_eq!(bits, vec![0, 0, 0, 0], "Should overflow to 0");
}

#[test]
fn test_counter_reset() {
    let mut state = CounterState::new(4);

    // Count to 3
    for _ in 0..3 {
        counter(0, 0, &mut state); // clock low, reset=0
        counter(1, 0, &mut state); // clock high, reset=0
    }
    assert_eq!(state.value, vec![1, 1, 0, 0]); // decimal 3

    // Reset: clock low with reset=1, then clock high with reset=1
    counter(0, 1, &mut state); // clock=0, reset=1
    let bits = counter(1, 1, &mut state); // clock=1, reset=1
    assert_eq!(bits, vec![0, 0, 0, 0], "Reset should zero the counter");
}

#[test]
fn test_counter_8bit_counts() {
    let mut state = CounterState::new(8);

    // Count to 5
    for _ in 0..5 {
        counter(0, 0, &mut state); // clock low
        counter(1, 0, &mut state); // clock high
    }

    // 5 in binary: 00000101 (LSB first: [1, 0, 1, 0, 0, 0, 0, 0])
    assert_eq!(state.value, vec![1, 0, 1, 0, 0, 0, 0, 0]);
}

File: code/packages/typescript/logic-gates/tests/sequential.test.ts
/**
 * Tests for sequential logic — latches, flip-flops, registers, and counters.
 *
 * These tests verify each component's truth table, state-holding behavior,
 * and edge cases. We test from the bottom up, since higher-level components
 * (like registers) depend on lower-level ones (like flip-flops).
 */

import { describe, it, expect } from "vitest";
import {
  srLatch,
  dLatch,
  dFlipFlop,
  register,
  shiftRegister,
  counter,
  type Bit,
  type FlipFlopState,
  type CounterState,
} from "../src/index.js";

// ===========================================================================
// SR LATCH TESTS
// ===========================================================================

describe("SR Latch", () => {
  it("set: S=1, R=0 should set Q=1", () => {
    const [q, qBar] = srLatch(1, 0);
    expect(q).toBe(1);
    expect(qBar).toBe(0);
  });

  it("reset: S=0, R=1 should reset Q=0", () => {
    const [q, qBar] = srLatch(0, 1);
    expect(q).toBe(0);
    expect(qBar).toBe(1);
  });

  it("hold after set: S=0, R=0 should hold Q=1", () => {
    let [q, qBar] = srLatch(1, 0);
    expect(q).toBe(1);
    // Now hold — pass the current state back
    [q, qBar] = srLatch(0, 0, q, qBar);
    expect(q).toBe(1);
    expect(qBar).toBe(0);
  });

  it("hold after reset: S=0, R=0 should hold Q=0", () => {
    let [q, qBar] = srLatch(0, 1);
    expect(q).toBe(0);
    [q, qBar] = srLatch(0, 0, q, qBar);
    expect(q).toBe(0);
    expect(qBar).toBe(1);
  });

  it("hold default state: S=0, R=0 should hold Q=0, Q_bar=1", () => {
    const [q, qBar] = srLatch(0, 0);
    expect(q).toBe(0);
    expect(qBar).toBe(1);
  });

  it("invalid both set and reset: S=1, R=1 gives both 0", () => {
    const [q, qBar] = srLatch(1, 1);
    expect(q).toBe(0);
    expect(qBar).toBe(0);
  });

  it("set then reset then set transitions correctly", () => {
    let [q, qBar] = srLatch(1, 0);
    expect(q).toBe(1);

    [q, qBar] = srLatch(0, 1, q, qBar);
    expect(q).toBe(0);
    expect(qBar).toBe(1);

    [q, qBar] = srLatch(1, 0, q, qBar);
    expect(q).toBe(1);
    expect(qBar).toBe(0);
  });

  it("multiple holds should be stable", () => {
    let [q, qBar] = srLatch(1, 0); // Set
    for (let i = 0; i < 10; i++) {
      [q, qBar] = srLatch(0, 0, q, qBar);
    }
    expect(q).toBe(1);
    expect(qBar).toBe(0);
  });

  it("non-number input throws TypeError", () => {
    expect(() => srLatch(true as unknown as Bit, 0)).toThrow(TypeError);
  });

  it("out-of-range input throws RangeError", () => {
    expect(() => srLatch(2 as Bit, 0)).toThrow(RangeError);
  });
});

// ===========================================================================
// D LATCH TESTS
// ===========================================================================

describe("D Latch", () => {
  it("store 1 when enabled", () => {
    const [q, qBar] = dLatch(1, 1);
    expect(q).toBe(1);
    expect(qBar).toBe(0);
  });

  it("store 0 when enabled", () => {
    const [q, qBar] = dLatch(0, 1);
    expect(q).toBe(0);
    expect(qBar).toBe(1);
  });

  it("hold when disabled after set", () => {
    let [q, qBar] = dLatch(1, 1); // Store 1
    expect(q).toBe(1);
    [q, qBar] = dLatch(0, 0, q, qBar); // Disable — data is 0 but ignored
    expect(q).toBe(1);
    expect(qBar).toBe(0);
  });

  it("hold when disabled after reset", () => {
    let [q, qBar] = dLatch(0, 1); // Store 0
    [q, qBar] = dLatch(1, 0, q, qBar); // Disable — data is 1 but ignored
    expect(q).toBe(0);
    expect(qBar).toBe(1);
  });

  it("transparent mode: output follows data when enabled", () => {
    let [q, qBar] = dLatch(1, 1);
    expect(q).toBe(1);
    [q, qBar] = dLatch(0, 1, q, qBar);
    expect(q).toBe(0);
    [q, qBar] = dLatch(1, 1, q, qBar);
    expect(q).toBe(1);
  });

  it("hold default state: Enable=0 holds Q=0", () => {
    const [q, qBar] = dLatch(1, 0);
    expect(q).toBe(0);
    expect(qBar).toBe(1);
  });

  it("non-number input throws TypeError", () => {
    expect(() => dLatch(true as unknown as Bit, 1)).toThrow(TypeError);
  });

  it("out-of-range input throws RangeError", () => {
    expect(() => dLatch(0, 2 as Bit)).toThrow(RangeError);
  });
});

// ===========================================================================
// D FLIP-FLOP TESTS
// ===========================================================================

describe("D Flip-Flop", () => {
  it("captures data on rising edge", () => {
    // Clock low: master absorbs data=1
    let [q, qBar, state] = dFlipFlop(1, 0);
    // Clock high: slave outputs master's stored value
    [q, qBar, state] = dFlipFlop(1, 1, state);
    expect(q).toBe(1);
    expect(qBar).toBe(0);
  });

  it("captures zero", () => {
    let [q, qBar, state] = dFlipFlop(0, 0);
    [q, qBar, state] = dFlipFlop(0, 1, state);
    expect(q).toBe(0);
    expect(qBar).toBe(1);
  });

  it("holds during clock high", () => {
    // Capture 1
    let [q, qBar, state] = dFlipFlop(1, 0);
    [q, qBar, state] = dFlipFlop(1, 1, state);
    expect(q).toBe(1);
    // Data changes to 0 while clock is still high — output should hold
    [q, qBar, state] = dFlipFlop(0, 1, state);
    expect(q).toBe(1);
  });

  it("new value on next edge", () => {
    // Capture 1
    let [q, qBar, state] = dFlipFlop(1, 0);
    [q, qBar, state] = dFlipFlop(1, 1, state);
    expect(q).toBe(1);

    // Next cycle: capture 0
    [q, qBar, state] = dFlipFlop(0, 0, state);
    [q, qBar, state] = dFlipFlop(0, 1, state);
    expect(q).toBe(0);
    expect(qBar).toBe(1);
  });

  it("multiple clock cycles with different values", () => {
    let state: FlipFlopState = {
      masterQ: 0,
      masterQBar: 1,
      slaveQ: 0,
      slaveQBar: 1,
    };
    const values: Bit[] = [1, 0, 1, 1, 0];
    for (const val of values) {
      let q: Bit;
      [, , state] = dFlipFlop(val, 0, state);
      [q, , state] = dFlipFlop(val, 1, state);
      expect(q).toBe(val);
    }
  });

  it("internal state has expected keys", () => {
    const [, , state] = dFlipFlop(1, 0);
    expect(state).toHaveProperty("masterQ");
    expect(state).toHaveProperty("masterQBar");
    expect(state).toHaveProperty("slaveQ");
    expect(state).toHaveProperty("slaveQBar");
  });

  it("non-number input throws TypeError", () => {
    expect(() => dFlipFlop(true as unknown as Bit, 0)).toThrow(TypeError);
  });

  it("out-of-range input throws RangeError", () => {
    expect(() => dFlipFlop(0, 2 as Bit)).toThrow(RangeError);
  });
});

// ===========================================================================
// REGISTER TESTS
// ===========================================================================

describe("Register", () => {
  it("stores a 4-bit value", () => {
    // Clock low: absorb data
    let [out, state] = register([1, 0, 1, 1], 0);
    // Clock high: output stored data
    [out, state] = register([1, 0, 1, 1], 1, state);
    expect(out).toEqual([1, 0, 1, 1]);
  });

  it("stores an 8-bit value", () => {
    const data: Bit[] = [1, 0, 0, 1, 1, 0, 1, 0];
    let [out, state] = register(data, 0);
    [out, state] = register(data, 1, state);
    expect(out).toEqual(data);
  });

  it("holds previous value then captures new data", () => {
    const data1: Bit[] = [1, 1, 0, 0];
    let [out, state] = register(data1, 0);
    [out, state] = register(data1, 1, state);
    expect(out).toEqual(data1);

    // Present new data, clock low then high
    const data2: Bit[] = [0, 0, 1, 1];
    [out, state] = register(data2, 0, state);
    [out, state] = register(data2, 1, state);
    expect(out).toEqual(data2);
  });

  it("width parameter enforces data length", () => {
    const [out] = register([1, 0], 0, undefined, 2);
    expect(out).toHaveLength(2);
  });

  it("width mismatch throws RangeError", () => {
    expect(() => register([1, 0, 1], 0, undefined, 2)).toThrow(
      "does not match width",
    );
  });

  it("empty data throws RangeError", () => {
    expect(() => register([], 0)).toThrow("must not be empty");
  });

  it("non-array data throws TypeError", () => {
    expect(() => register(1 as unknown as Bit[], 0)).toThrow(TypeError);
  });

  it("state length mismatch throws RangeError", () => {
    const [, state] = register([1, 0], 0);
    expect(() => register([1, 0, 1], 0, state)).toThrow(
      "does not match data length",
    );
  });

  it("single bit register", () => {
    let [out, state] = register([1], 0);
    [out, state] = register([1], 1, state);
    expect(out).toEqual([1]);
  });

  it("invalid bit in data throws", () => {
    expect(() => register([0, 2 as Bit, 1], 0)).toThrow();
  });

  it("all zeros", () => {
    const data: Bit[] = [0, 0, 0, 0];
    let [out, state] = register(data, 0);
    [out, state] = register(data, 1, state);
    expect(out).toEqual([0, 0, 0, 0]);
  });

  it("all ones", () => {
    const data: Bit[] = [1, 1, 1, 1];
    let [out, state] = register(data, 0);
    [out, state] = register(data, 1, state);
    expect(out).toEqual([1, 1, 1, 1]);
  });
});

// ===========================================================================
// SHIFT REGISTER TESTS
// ===========================================================================

describe("Shift Register", () => {
  it("shift right single bit", () => {
    let [out, sout, state] = shiftRegister(1, 0, undefined, 4);
    [out, sout, state] = shiftRegister(1, 1, state, 4);
    expect(out).toEqual([1, 0, 0, 0]);
    expect(sout).toBe(0); // Nothing shifted out yet
  });

  it("shift right fill", () => {
    let state: FlipFlopState[] | undefined;
    let out: Bit[];
    for (let i = 0; i < 4; i++) {
      [, , state] = shiftRegister(1, 0, state, 4);
      [out, , state] = shiftRegister(1, 1, state, 4);
    }
    expect(out!).toEqual([1, 1, 1, 1]);
  });

  it("shift right serial out", () => {
    // Fill with 1s
    let state: FlipFlopState[] | undefined;
    let sout: Bit;
    let out: Bit[];
    for (let i = 0; i < 4; i++) {
      [, , state] = shiftRegister(1, 0, state, 4);
      [, sout, state] = shiftRegister(1, 1, state, 4);
    }

    // Now shift in 0 — the rightmost 1 should come out
    [, , state] = shiftRegister(0, 0, state, 4);
    [out, sout, state] = shiftRegister(0, 1, state, 4);
    expect(sout).toBe(1);
    expect(out).toEqual([0, 1, 1, 1]);
  });

  it("shift left single bit", () => {
    let [out, sout, state] = shiftRegister(1, 0, undefined, 4, "left");
    [out, sout, state] = shiftRegister(1, 1, state, 4, "left");
    expect(out).toEqual([0, 0, 0, 1]);
    expect(sout).toBe(0);
  });

  it("shift left fill", () => {
    let state: FlipFlopState[] | undefined;
    let out: Bit[];
    for (let i = 0; i < 4; i++) {
      [, , state] = shiftRegister(1, 0, state, 4, "left");
      [out, , state] = shiftRegister(1, 1, state, 4, "left");
    }
    expect(out!).toEqual([1, 1, 1, 1]);
  });

  it("shift left serial out", () => {
    // Fill with 1s
    let state: FlipFlopState[] | undefined;
    let sout: Bit;
    let out: Bit[];
    for (let i = 0; i < 4; i++) {
      [, , state] = shiftRegister(1, 0, state, 4, "left");
      [, , state] = shiftRegister(1, 1, state, 4, "left");
    }

    // Shift in 0 — leftmost 1 should come out
    [, , state] = shiftRegister(0, 0, state, 4, "left");
    [out, sout, state] = shiftRegister(0, 1, state, 4, "left");
    expect(sout).toBe(1);
    expect(out).toEqual([1, 1, 1, 0]);
  });

  it("pattern shift right", () => {
    const pattern: Bit[] = [1, 0, 1, 0];
    let state: FlipFlopState[] | undefined;
    let out: Bit[];
    for (const bit of pattern) {
      [, , state] = shiftRegister(bit, 0, state, 4);
      [out, , state] = shiftRegister(bit, 1, state, 4);
    }
    // Last bit shifted in is at index 0, first bit at index 3
    expect(out!).toEqual([0, 1, 0, 1]);
  });

  it("width=1", () => {
    let [out, sout, state] = shiftRegister(1, 0, undefined, 1);
    [out, sout, state] = shiftRegister(1, 1, state, 1);
    expect(out).toEqual([1]);
    expect(sout).toBe(0);
  });

  it("invalid direction throws", () => {
    expect(() =>
      shiftRegister(1, 0, undefined, 4, "up" as "right"),
    ).toThrow("direction must be");
  });

  it("invalid width throws", () => {
    expect(() => shiftRegister(1, 0, undefined, 0)).toThrow("width must be");
  });

  it("state length mismatch throws", () => {
    const [, , state] = shiftRegister(1, 0, undefined, 4);
    expect(() => shiftRegister(1, 0, state, 3)).toThrow(
      "does not match width",
    );
  });

  it("default width is 8", () => {
    const [out] = shiftRegister(0, 0);
    expect(out).toHaveLength(8);
  });
});

// ===========================================================================
// COUNTER TESTS
// ===========================================================================

describe("Counter", () => {
  it("counts to three", () => {
    /**
     * Counter should count 1, 2, 3, 4 after successive clock edges.
     *
     * The counter increments on each rising clock edge, so the first
     * tick produces 1 (not 0). This matches hardware behavior — a real
     * counter starts at 0 and the first edge gives 1.
     */
    let state: CounterState | undefined;
    const expectedValues: Bit[][] = [
      [1, 0, 0, 0], // 1 (first tick)
      [0, 1, 0, 0], // 2
      [1, 1, 0, 0], // 3
      [0, 0, 1, 0], // 4
    ];

    for (let i = 0; i < 4; i++) {
      let bits: Bit[];
      [bits, state] = counter(0, 0, state, 4);
      [bits, state] = counter(1, 0, state, 4);
      expect(bits).toEqual(expectedValues[i]);
    }
  });

  it("reset forces counter to zero", () => {
    let state: CounterState | undefined;
    // Count up a couple times
    for (let i = 0; i < 3; i++) {
      [, state] = counter(0, 0, state, 4);
      [, state] = counter(1, 0, state, 4);
    }

    // Reset
    let bits: Bit[];
    [bits, state] = counter(0, 1, state, 4);
    [bits, state] = counter(1, 1, state, 4);
    expect(bits).toEqual([0, 0, 0, 0]);
  });

  it("counts normally after reset", () => {
    let state: CounterState | undefined;
    // Count to 3
    for (let i = 0; i < 3; i++) {
      [, state] = counter(0, 0, state, 4);
      [, state] = counter(1, 0, state, 4);
    }

    // Reset
    [, state] = counter(0, 1, state, 4);
    [, state] = counter(1, 1, state, 4);

    // Count again
    let bits: Bit[];
    [bits, state] = counter(0, 0, state, 4);
    [bits, state] = counter(1, 0, state, 4);
    expect(bits).toEqual([1, 0, 0, 0]); // 1
  });

  it("overflow wraps to zero", () => {
    let state: CounterState | undefined;
    const width = 3; // Max value = 7

    // Count to 7 (7 ticks since each tick increments)
    let bits: Bit[];
    for (let i = 0; i < 7; i++) {
      [, state] = counter(0, 0, state, width);
      [bits, state] = counter(1, 0, state, width);
    }
    expect(bits!).toEqual([1, 1, 1]); // 7

    // One more tick should wrap to 0
    [, state] = counter(0, 0, state, width);
    [bits, state] = counter(1, 0, state, width);
    expect(bits).toEqual([0, 0, 0]); // 0 (overflow)
  });

  it("1-bit counter toggles", () => {
    let state: CounterState | undefined;
    let bits: Bit[];

    // Tick 1 — starts at 0, increments to 1
    [, state] = counter(0, 0, state, 1);
    [bits, state] = counter(1, 0, state, 1);
    expect(bits).toEqual([1]);

    // Tick 2 — wraps to 0
    [, state] = counter(0, 0, state, 1);
    [bits, state] = counter(1, 0, state, 1);
    expect(bits).toEqual([0]);

    // Tick 3 — back to 1
    [, state] = counter(0, 0, state, 1);
    [bits, state] = counter(1, 0, state, 1);
    expect(bits).toEqual([1]);
  });

  it("invalid width throws", () => {
    expect(() => counter(0, 0, undefined, 0)).toThrow("width must be");
  });

  it("default width is 8", () => {
    const [bits] = counter(0);
    expect(bits).toHaveLength(8);
  });

  it("state has value and ffState keys", () => {
    const [, state] = counter(0, 0, undefined, 4);
    expect(state).toHaveProperty("value");
    expect(state).toHaveProperty("ffState");
  });
});
